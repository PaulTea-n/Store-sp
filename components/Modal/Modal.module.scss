.modalContainer {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    transform: translate(0, 0);
    z-index: 1000;
    display: flex;
    justify-content: center;
    align-items: center;
}


.backgroundModal {
    width: 100%;
    height: 100vh;
    position: absolute;
    top: 0;
    left: 0;
    background: rgba(0, 0, 0, 0.8);
    z-index: 1;
    cursor: pointer;
}

.modal {
    // position: fixed;
    position: absolute;
    border-radius: 10px;
    width: 60%;
    left: 20%;
    top: 15%;
    border: 1px solid black;
    background-color: #F33B0F;
    color: #FFF;
    z-index: 100;
}

.closeButton {
    width: 40px;
    height: 40px;
    border-radius: 10px;
    border: none;
    cursor: pointer;
    font-size: 22px;
    color: #FFF;
    background-color: #9E4439;
}

.header {
    display: flex;
    justify-content: space-between;
    background-color: #D95545;
    border-radius: 10px;
}

.headerText {
    padding-left: 20px;
    margin: auto;
}

.textModal {
    text-align: center;
    // padding: 20px;
    padding-top: 0px;
}

.btnContainer {

    text-align: center;

    .btnModal {
        width: 600px;
    }

}