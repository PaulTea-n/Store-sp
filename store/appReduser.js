import { combineReducers } from "redux";
import cartReducer from "./cart/cartReducer";
import favoriteReducer from "./favorite/favoriteReducer";
import modalReduser from "./modal/modalReducer";
import productsReduser from "./products/productsReducer";

const appReduser = combineReducers({
  modal: modalReduser,
  products: productsReduser,
  cart: cartReducer,
  favorite: favoriteReducer,
});

export default appReduser;
