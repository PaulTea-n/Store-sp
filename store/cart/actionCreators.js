import {
  ADD_TO_CART,
  GET_CART_LOCAL_STORAGE,
  INCREMENT_CART_ITEM,
  DICREMENT_CART_ITEM,
  DELETE_CART_ITEM,
  GET_CART_COUNTER_LOCAL_STORAGE,
  CLEAN_PRODUCTS_TO_BUY,
} from "./actions";

export const addToCartAC = (payload) => {
  return { type: ADD_TO_CART, payload };
};

export const getCartLocalStorageAC = (payload) => {
  return { type: GET_CART_LOCAL_STORAGE, payload };
};

export const incrementCartItemAC = (payload) => {
  return { type: INCREMENT_CART_ITEM, payload };
};

export const dicrementCartItemAC = (payload) => {
  return { type: DICREMENT_CART_ITEM, payload };
};

export const deleteCartItemAC = (payload) => {
  return { type: DELETE_CART_ITEM, payload };
};
export const getCartCounterLocalStorageAC = (payload) => {
  return { type: GET_CART_COUNTER_LOCAL_STORAGE, payload };
};

export const cleanProductsToBuy = (payload) => {
  return { type: CLEAN_PRODUCTS_TO_BUY, payload };
};
