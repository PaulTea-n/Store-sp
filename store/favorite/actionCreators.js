import {
  ADD_TO_FAVORITE,
  GET_FAVORITE_LOCAL_STORAGE,
  FAVORITE_CAUNTER,
} from "./actions";

export const addToFavoriteAC = (payload) => {
  return { type: ADD_TO_FAVORITE, payload };
};

export const getFavoriteLocalStorageAC = () => async (dispatch) => {
  const favorites = JSON.parse(localStorage.getItem("favorites"));
  const favoriteCounter = JSON.parse(localStorage.getItem("favoriteCounter"));

  if (favorites !== null) {
    dispatch({ type: GET_FAVORITE_LOCAL_STORAGE, payload: favorites });
  }
  if (favoriteCounter !== null) {
    dispatch({
      type: FAVORITE_CAUNTER,
      payload: favoriteCounter,
    });
  }
};

export const favoriteCaunterAC = (payload) => {
  return { type: FAVORITE_CAUNTER, payload };
};
