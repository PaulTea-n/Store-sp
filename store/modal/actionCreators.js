import { OPEN_MODAL, CLOSE_MODAL, SET_MODAL_DATA } from "./actions";

export const openModalAC = () => ({ type: OPEN_MODAL });
export const closeModalAC = () => ({ type: CLOSE_MODAL });
export const setModalDataAC = (payload) => ({ type: SET_MODAL_DATA, payload });
