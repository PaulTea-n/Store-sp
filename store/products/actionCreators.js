import {
  GET_PRODUCTS,
  GET_PRODUCTS_LOCAL_STORAGE,
  SET_FAVORITE_PRODUCTS,
} from "./actions";

export const getProductsAC = () => async (dispatch) => {
  const products = await fetch("./products.json").then((res) => res.json());
  dispatch({ type: GET_PRODUCTS, payload: products });
};

export const getProductsLocalStorageAC = (payload) => {
  return { type: GET_PRODUCTS_LOCAL_STORAGE, payload };
};

export const setFavoriteProductsAC = (payload) => {
  return { type: SET_FAVORITE_PRODUCTS, payload };
};
