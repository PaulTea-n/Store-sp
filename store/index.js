import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import appReduser from "./appReduser";
import thunk from "redux-thunk";

const store = createStore(
  appReduser,
  composeWithDevTools(applyMiddleware(thunk))
);

export default store;
