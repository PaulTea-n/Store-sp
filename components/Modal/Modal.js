import React from "react";
import styles from "./Modal.module.scss";
import PropTypes from "prop-types";
import Button from "../Button/Button";
import { useDispatch, useSelector } from "react-redux";
import { closeModalAC } from "../../store/modal/actionCreators";

const Modal = (props) => {
  const { confirmBtnAction, text } = props;
  const isOpenModalFirst = useSelector((store) => store.modal.isOpenModalFirst);
  const modalData = useSelector((store) => store.modal.modalData);
  const dispatch = useDispatch();

  const closeModal = () => {
    dispatch(closeModalAC());
  };

  if (!isOpenModalFirst) {
    return null;
  }

  return (
    <div data-testid="modal-container" className={styles.modalContainer}>
      <div className={styles.backgroundModal} onClick={closeModal}></div>
      <div className={styles.modal}>
        <div className={styles.header}>
          <h1 className={styles.headerText}>{modalData.name}</h1>
          <button className={styles.closeButton} handleClick={closeModal}>
            X
          </button>
        </div>
        <div>
          <h2 className={styles.textModal}>{text}</h2>
          <div className={styles.btnContainer}>
            <Button
              className={styles.btnModal}
              text="НІ"
              handleClick={() => {
                closeModal();
              }}
              backgroundColor="#7A7A7A"
            />
            <Button
              className={styles.btnModal}
              handleClick={() => {
                confirmBtnAction();
                closeModal();
              }}
              text="ТАК"
              backgroundColor="#001542"
            />
          </div>
        </div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  confirmBtnAction: PropTypes.func,
  text: PropTypes.string,
};

Modal.defaultProps = {
  text: "Add to cart",
  confirmBtnAction: () => {},
};

export default Modal;
