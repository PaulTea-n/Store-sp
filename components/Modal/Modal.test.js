import Modal from "./Modal";
import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider, useDispatch } from "react-redux";
import store from "../../store";
import { openModalAC, closeModalAC } from "../../store/modal/actionCreators";

const confirmBtnAction = jest.fn();

const Component = ({ text, confirmBtnAction }) => {
  const dispatch = useDispatch();

  return (
    <>
      <Modal text={text} confirmBtnAction={confirmBtnAction} />
      <button
        onClick={() => {
          dispatch(openModalAC());
        }}
      >
        OPEN
      </button>
      <button
        onClick={() => {
          dispatch(closeModalAC());
        }}
      >
        CLOSE
      </button>
    </>
  );
};

const MockedProvider = ({ text, confirmBtnAction }) => {
  return (
    <Provider store={store}>
      <Component text={text} confirmBtnAction={confirmBtnAction} />
    </Provider>
  );
};

describe("Modal renders", () => {
  test("should Modal no render without isOpenModalFirst props", () => {
    const { asFragment } = render(<MockedProvider text="Add to cart" />);
    fireEvent.click(screen.getByText("OPEN"));
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Modal confirmBtnAction works", () => {
  test("should confirmBtnAction be called when 'ТАК' btn is clicked", () => {
    render(
      <MockedProvider text="Add to cart" confirmBtnAction={confirmBtnAction} />
    );

    fireEvent.click(screen.getByText("OPEN"));
    fireEvent.click(screen.getByText("ТАК"));

    expect(confirmBtnAction).toHaveBeenCalled();
  });
});

describe("Modal opens and closes on state changes", () => {
  test("should Modal open when isOpenModalFirst is true", () => {
    render(<MockedProvider text="test text" />, { store });

    fireEvent.click(screen.getByText("OPEN"));

    expect(screen.getByTestId("modal-container")).toBeInTheDocument();
  });

  test("should Modal close when isOpenModalFirst is false", () => {
    render(<MockedProvider text="test text" />, { store });

    fireEvent.click(screen.getByText("OPEN"));
    fireEvent.click(screen.getByText("CLOSE"));

    expect(screen.queryByTestId("modal-container")).not.toBeInTheDocument();
  });
});

describe("Modal closeModal fn works", () => {
  test("should closeModal be called when bg is clicked", () => {
    render(<MockedProvider text="Add to cart" />);

    fireEvent.click(screen.getByText("OPEN"));

    expect(screen.getByTestId("modal-container")).toBeInTheDocument();
  });

  test("should closeModal be called when X is clicked", () => {
    render(<MockedProvider text="Add to cart" />);

    fireEvent.click(screen.getByText("OPEN"));
    fireEvent.click(screen.getByText("X"));

    expect(screen.getByTestId("modal-container")).toBeInTheDocument();
  });

  test("should closeModal be called when 'НІ' is clicked", () => {
    render(
      <MockedProvider text="Add to cart" confirmBtnAction={confirmBtnAction} />
    );

    fireEvent.click(screen.getByText("OPEN"));
    fireEvent.click(screen.getByText("НІ"));

    expect(screen.queryByTestId("modal-container")).not.toBeInTheDocument();
  });
});
