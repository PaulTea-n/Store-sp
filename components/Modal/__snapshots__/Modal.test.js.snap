// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders should Modal no render without isOpenModalFirst props 1`] = `
<DocumentFragment>
  <div
    class="modalContainer"
    data-testid="modal-container"
  >
    <div
      class="backgroundModal"
    />
    <div
      class="modal"
    >
      <div
        class="header"
      >
        <h1
          class="headerText"
        />
        <button
          class="closeButton"
        >
          X
        </button>
      </div>
      <div>
        <h2
          class="textModal"
        >
          Add to cart
        </h2>
        <div
          class="btnContainer"
        >
          <button
            class="btn"
            style="background-color: rgb(122, 122, 122);"
            type="button"
          >
            НІ
          </button>
          <button
            class="btn"
            style="background-color: rgb(0, 21, 66);"
            type="button"
          >
            ТАК
          </button>
        </div>
      </div>
    </div>
  </div>
  <button>
    OPEN
  </button>
  <button>
    CLOSE
  </button>
</DocumentFragment>
`;
